### DISCUSSION TITLE
Finding Off-Heap leak but profiling is inconclusive

### DISCUSSION THREAD

**Discussion by Lonzak** [async-profiler/async-profiler#491] [Q&A] [answered]
I am trying to find a memory leak since I get an `java.lang.OutOfMemoryError: Direct buffer memory` after upgrading the application server (wildfly 13 to 18). The whole story is described [here](https://stackoverflow.com/questions/69769819/direct-buffer-memory-outofmemoryerror-after-java-wildfly-update).
I watched the [video ](https://www.youtube.com/watch?v=c755fFv1Rnk) and tried allocation profiling with the following commands:

` ./profiler.sh -d 10 -e malloc -f malloc.html jps`

I tried to vary the times (30,300 ...) however the flame graphs generated are not conclusive. I also tried to profile mprotect

` ./profiler.sh -d 300 -e mprotect -f mprotect.html jps`

But no matter how long I wait the result is always empty. Does anybody know the reason for that?

**Comment by krzysztofslusarski** [comment]
If you want to know which part of a code creates DirectByteBuffers you can use profiler in method mode for ```java.nio.ByteBuffer.allocateDirect``` or ```java.nio.DirectByteBuffer.<init>``` or native method ```Unsafe_AllocateMemory0``` (you can see explanation in README in which situation profiling java method is not safe).

If you want to know why those objects are not removed by GC you need to look at "path to gc" of those objects in heap dump analyzer. This will show you why those object are alive, then you need to find out by yourself why such a situation occurs.

**Comment by krzysztofslusarski** [comment]
The other thing you can profile is ```page-faults``` event. This mode can suggest when the RSS of your process is growing.

**Comment by apangin** [comment]
Please take a look at the related discussion: #453

Not all native allocators use `mprotect`, some just go directly to `mmap`.

I recently added a special profiling mode `-e nativemem` for finding native memory leaks (see [README](https://github.com/jvm-profiling-tools/async-profiler/blob/malloc/README.md#finding-native-memory-leaks)).
It's available in the [malloc](https://github.com/jvm-profiling-tools/async-profiler/tree/malloc) branch.
The prebuilt binaries are here: https://profiler.tools/builds/async-profiler-2.1-malloc-linux-x64.tar.gz

Since you already know that the problem is related to Direct ByteBuffers, Krzysztof's suggestion to profile `Unsafe_AllocateMemory0` should also help.

**Reply by Lonzak** [reply]
I downloaded the file and used the -e nativemem option however got the following error:
[ERROR] Must LD_PRELOAD profiler library for native memory profiling
So I used:
export LD_PRELOAD=build/libasyncProfiler.so
But still no luck - any hints how I can make it work?

**Reply by apangin** [reply]
Have you seen ["Finding native memory leaks"](https://github.com/jvm-profiling-tools/async-profiler/blob/malloc/README.md#finding-native-memory-leaks) chapter in README?

LD_PRELOAD needs to be set for Java process, not for profiler.sh.

**Reply by Lonzak** [reply]
Ok I read the readme. Now I use 
`LD_PRELOAD=build/libasyncProfiler.so java -agentlib:asyncProfiler=start,event=nativemem,file=malloc.jfr jps`
to start profiling however get the following mesage:

```
Error occurred during initialization of VM
Could not find agent library asyncProfiler on the library path, with error: libasyncProfiler.so: cannot open shared object file: No such file or directory
```

But the does exists in build/libasyncProfiler.so. Any idea what can cause this?

**Reply by cloud-66** [reply]
Hello, Is there any plans for adding -e nativemem  to master branch ?

**Reply by apangin** [reply]
@cloud-66 Answered [here](https://github.com/async-profiler/async-profiler/discussions/836#discussioncomment-10055271)

**Reply by vpriscan** [reply]
Hello @apangin, what is the current status of _-e nativemem_ ?

**Reply by apangin** [reply]
@vpriscan Native memory profiling feature has been [integrated](https://github.com/async-profiler/async-profiler/pull/1064). It's already available in [nightly builds](https://github.com/async-profiler/async-profiler/releases/tag/nightly).

**Comment by krzysztofslusarski** [comment]
Can you change ```agentlib:asyncProfiler``` to ```agentlib:/full/path/to/libasyncProfiler.so```?

**Reply by Lonzak** [reply]
I tried:
`LD_PRELOAD=build java -agentlib:build/asyncProfiler.so=start,event=nativemem,file=malloc.jfr jps`
Result:
```
ERROR: ld.so: object 'build' from LD_PRELOAD cannot be preloaded (cannot open shared object file): ignored.
Error occurred during initialization of VM
Could not find agent library build/asyncProfiler.so on the library path, with error: libbuild/asyncProfiler.so.so: cannot open shared object file: No such file or directory
```
So `.so.so` are too many. So without `.so`:
` LD_PRELOAD=build java -agentlib:build/asyncProfiler=start,event=nativemem,file=malloc.jfr jps`
Result:
```
ERROR: ld.so: object 'build' from LD_PRELOAD cannot be preloaded (cannot open shared object file): ignored.
Error occurred during initialization of VM
Could not find agent library build/asyncProfiler on the library path, with error: libbuild/asyncProfiler.so: cannot open shared object file: No such file or directory
```
Next try:
`LD_PRELOAD=build/libasyncProfiler.so java -agentlib:/full/path/async-profiler-2.1-malloc-linux-x64/build/libasyncProfiler=start,event=nativemem,file=malloc.jfr jps`

```
Error occurred during initialization of VM
Could not find agent library /full/path/async-profiler-2.1-malloc-linux-x64/build/libasyncProfiler on the library path, with error: lib/full/path/async-profiler-2.1-malloc-linux-x64/build/libasyncProfiler.so: cannot open shared object file: No such file or directory
```

**Reply by apangin** [reply]
The options is either `-agentlib:asyncProfiler=...` or `-agentpath:/full/path/libasyncProfiler.so=...`

- `-agentlib` with the short library name (if .so file is on the standard library path like `/usr/lib64`);
- `-agentpath` with the full path name, inlcuding `.so` extension.

**Reply by Lonzak** [reply]
I don't get it to run.

`LD_PRELOAD=/full/path/async-profiler-2.1-malloc-linux-x64/build/libasyncProfiler.so java -agentlib:asyncProfiler=start,event=nativemem,file=malloc.jfr jps`

The error is:
```
Error occurred during initialization of VM
Could not find agent library asyncProfiler on the library path, with error: libasyncProfiler.so: cannot open shared object file: No such file or directory
```
But the file path is definitely correct (autocompleted file path). The fie properties are also set correctly (+x etc). Or does it need root rights to be executed?

**Comment by krzysztofslusarski** [comment]
How about:
```LD_PRELOAD=/full/path/async-profiler-2.1-malloc-linux-x64/build/libasyncProfiler.so java -agentlib:/full/path/async-profiler-2.1-malloc-linux-x64/build/libasyncProfiler.so=start,event=nativemem,file=malloc.jfr jps``` ?

**Reply by Lonzak** [reply]
Nearly ;-) as Andrei had written: `-agentpath with the full path name, inlcuding .so extension.` So I used agentpath instead of agentlib:

`LD_PRELOAD=/full/path/async-profiler-2.1-malloc-linux-x64/build/libasyncProfiler.so java -agentpath:/full/path/async-profiler-2.1-malloc-linux-x64/build/libasyncProfiler.so=start,event=nativemem,file=malloc.jfr jps`

Then I see: `Profiling started` which is one step further 👍 
However then there is a:

```
Error: main class could not be found or loaded
Reason: java.lang.ClassNotFoundException: jps
```
So there seem to be some syntax issue with the rest of the command. Any idea? I also tried the PID but then it just writes the PID instead of jps

**Reply by apangin** [reply]
Sorry if I was unclear. I meant that your application needs to be started with `LD_PRELOAD`, not the profiler.

I see that the current setup of the native leak profiler is not trivial, and I have ideas how to improve this in future. However, right now `nativemem` mode works only if the target process has been started with the LD_PRELOADED'ed library.

**Comment by Lonzak** [comment, marked as answer]
Thank you for your support and this great tool!

LD_PRELOAD=/full/path/async-profiler-2.1-malloc-linux-x64/build/libasyncProfiler.so java -agentpath:/full/path/async-profiler-2.1-malloc-linux-x64/build/libasyncProfiler.so=start,event=nativemem,file=malloc.jfr YourJavaApplication2BProfiled

**Reply by zhongyujiang** [reply]
Hi @Lonzak, I am using `nativemem` mode too, I followed your answer and sucessfully got the job started,  however I didn't see malloc.jfr file, and when I run `./profiler.sh stop <pid> ` to stop the job, I got an error  like "profiling not started".  I was wondering if this happened to you too, if so, could you please  share your experience of success? Thanks.

**Reply by Lonzak** [reply]
Hi,

no I didn not run into this problem. I would open another ticket and people here are quite helpful and you'll most certainly get an reply...

**Reply by zhongyujiang** [reply]
OK, thanks for your reply and advice!

**Reply by AnshitaJohri** [reply]
I am facing the same issue  "profiling not started", any advice will be very helpful

**Reply by apangin** [reply]
@AnshitaJohri "profiling not started" message means profiler failed to start. Check the VM logs (stdout of the target Java application) for the exact error message from the profiler - it should be somewhere in the beginning (during VM startup).

**Reply by AnshitaJohri** [reply]
Thanks apangin, i really appreciate for taking time and replying to the query , please see start up logs for my application, This is from my local I see below

[ERROR] Unsupported event type

~/Desktop/Workspace/broken/jemalloc$ LD_PRELOAD=/home/usr/Desktop/Workspace/broken/async-profiler-2.9-linux-x64/build/libasyncProfiler.so /home/usr/Desktop/softwares/idea-IU-221.5080.210/jbr/bin/java -agentpath:/home/usr/Desktop/Workspace/broken/async-profiler-2.9-linux-x64/build/libasyncProfiler.so=start,event=nativemem,file=malloc.html -jar /home/usr/Desktop/Workspace/branch-reporting-service/build/libs/branch-reporting-service.jar --spring.profiles.active=local
[ERROR] Unsupported event type

  .   ____          _            __ _ _
 /\\ / ___'_ __ _ _(_)_ __  __ _ \ \ \ \
( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
 \\/  ___)| |_)| | | | | || (_| |  ) ) ) )
  '  |____| .__|_| |_|_| |_\__, | / / / /
 =========|_|==============|___/=/_/_/_/
 :: Spring Boot ::                (v2.7.3)

2023-04-08 00:59:40.716  INFO 48986 --- [           main] .f.p.b.BranchReportingServiceApplication : Starting BranchReportingServiceApplication using Java 11.0.14.1 on f5407621-latitude-5520 with PID 48986 (/home/usr/Desktop/Workspace/branch-reporting-service/build/libs/branch-reporting-service.jar started by usr in /home/usr/Desktop/Workspace/broken/jemalloc)
2023-04-08 00:59:40.719 DEBUG 48986 --- [           main] .f.p.b.BranchReportingServiceApplication : Running with Spring Boot v2.7.3, Spring v5.3.22
2023-04-08 00:59:40.719  INFO 48986 --- [           main] .f.p.b.BranchReportingServiceApplication : The following 1 profile is active: "local"
2023-04-08 00:59:41.814  INFO 48986 --- [           main] .s.d.r.c.RepositoryConfigurationDelegate : Multiple Spring Data modules found, entering strict repository configuration mode
2023-04-08 00:59:41.816  INFO 48986 --- [           main] .s.d.r.c.RepositoryConfigurationDelegate : Bootstrapping Spring Data JPA repositories in DEFAULT mode.
2023-04-08 00:59:42.000  INFO 48986 --- [           main] .s.d.r.c.RepositoryConfigurationDelegate : Finished Spring Data repository scanning in 177 ms. Found 5 JPA repository interfaces.
2023-04-08 00:59:42.205  INFO 48986 --- [           main] o.s.cloud.context.scope.GenericScope     : BeanFactory id=e5ba4a04-a43e-3d44-bcd4-d30c018cebaf
2023-04-08 00:59:42.369  INFO 48986 --- [           main] trationDelegate$BeanPostProcessorChecker : Bean 'org.springframework.ws.config.annotation.DelegatingWsConfiguration' of type [org.springframework.ws.config.annotation.DelegatingWsConfiguration$$EnhancerBySpringCGLIB$$950f30fe] is not eligible for getting processed by all BeanPostProcessors (for example: not eligible for auto-proxying)
2023-04-08 00:59:42.416  INFO 48986 --- [           main] .w.s.a.s.AnnotationActionEndpointMapping : Supporting [WS-Addressing August 2004, WS-Addressing 1.0]
2023-04-08 00:59:42.757  INFO 48986 --- [           main] o.s.b.w.embedded.tomcat.TomcatWebServer  : Tomcat initialized with port(s): 8090 (http)
2023-04-08 00:59:42.767  INFO 48986 --- [           main] o.apache.catalina.core.StandardService   : Starting service [Tomcat]
2023-04-08 00:59:42.767  INFO 48986 --- [           main] org.apache.catalina.core.StandardEngine  : Starting Servlet engine: [Apache Tomcat/9.0.65]
2023-04-08 00:59:42.834  INFO 48986 --- [           main] o.a.c.c.C.[Tomcat].[localhost].[/]       : Initializing Spring embedded WebApplicationContext
2023-04-08 00:59:42.834  INFO 48986 --- [           main] w.s.c.ServletWebServerApplicationContext : Root WebApplicationContext: initialization completed in 2064 ms
2023-04-08 00:59:42.929 DEBUG 48986 --- [           main] com.zaxxer.hikari.HikariConfig           : Driver class com.mysql.cj.jdbc.Driver found in Thread context class loader TomcatEmbeddedWebappClassLoader
  context: ROOT
  delegate: true
----------> Parent Classloader:

**Reply by apangin** [reply]
`nativemem` profiling is not yet supported in the mainline async-profiler..
You need the special version of async-profiler from [`malloc`](https://github.com/async-profiler/async-profiler/tree/malloc) branch. You can build it yourself or take the prebuilt binaries at the location mentioned [here](https://github.com/async-profiler/async-profiler/discussions/491#discussioncomment-1590216).

**Reply by AnshitaJohri** [reply]
Thanks apangin, it works and i was able to generate the flame graph, but now i am not sure how to analyse them to identify the possible cause of leak , attached, in case you can please please direct me here
[native.zip](https://github.com/async-profiler/async-profiler/files/11186195/native.zip)

**Reply by apangin** [reply]
Why do you think there is a native memory leak?
(please open a new issue/discussion, as this seems not related to the previous topic)
